### PR TITLE
support cfg_vis on fields

### PR DESCRIPTION
**Pull request by TOETOE55**
cfg_vis on fields with great compile error message

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -99,5 +122,130 @@ pub fn cfg_vis(
...
+
+    let toks = cfg_vis_fields_impl(item)
+        .map(|item| quote! { #item })
+        .unwrap_or_else(|err| err.to_compile_error());
+
+    proc_macro::TokenStream::from(toks)
+}
+
+fn cfg_vis_fields_impl(mut item: syn::Item) -> syn::Result<syn::Item> {
+    let fields = match &mut item {
```
Comment (mu001999) [inline src/lib.rs]: Don't support enums?

File: src/lib.rs [hunk 2]
```diff
@@ -99,5 +122,130 @@ pub fn cfg_vis(
...
+    *fields = find_replace_cfg_vis(std::mem::take(fields))?;
+
+    Ok(item)
+}
+
+fn find_replace_cfg_vis(
+    fields: Punctuated<syn::Field, syn::Token![,]>,
+) -> syn::Result<Punctuated<syn::Field, syn::Token![,]>> {
+    let mut fields_replaced = Punctuated::new();
+    for mut field in fields {
```
Comment (mu001999) [inline src/lib.rs]: Why not use `fields.into_iter().flat_map(...).collect()` (

Comment (TOETOE55) [reply src/lib.rs]: because I want to early return from the loop.

File: src/lib.rs (hunk 1 above)
Comment (TOETOE55) [reply src/lib.rs]: Enum variant's fields don’t have visibility.

File: src/lib.rs (hunk 2 above)
Comment (mu001999) [reply src/lib.rs]: How about this within `flat_map`:
```rust
let pos = if let ... { ... } else {
    return vec![field];
}
```

File: src/lib.rs (hunk 1 above)
Comment (mu001999) [reply src/lib.rs]: ,,ԾㅂԾ,,

File: src/lib.rs (hunk 2 above)
Comment (TOETOE55) [reply src/lib.rs]: How do you deal with `guard_cfg_vis_unique(&field.attrs, false)?`

Comment (mu001999) [reply src/lib.rs]: sry, I overlooked `? `. how about the following:
```rust
fn find_replace_cfg_vis(
    fields: Punctuated<syn::Field, syn::Token![,]>,
) -> syn::Result<Punctuated<syn::Field, syn::Token![,]>> {
    fields
        .into_iter()
        .try_fold(Punctuated::new(), |mut fields, mut field| {
            if let Some(pos) = guard_cfg_vis_unique(&field.attrs, false)? {
                let attr = &field.attrs[pos].tokens;
                let CfgVisAttrArgsWithParens(CfgVisAttrArgs { cfg, vis }) = parse_quote!(#attr);

                let mut field_replaced = field.clone();
                field_replaced.attrs[pos] = parse_quote! { #[cfg(#cfg)] };
                field_replaced.vis = vis;
                fields.push(field_replaced);

                field.attrs[pos] = parse_quote! { #[cfg(not(#cfg))] };
            }
            fields.push(field);
            Ok(fields)
        })
}
```

Comment (TOETOE55) [reply src/lib.rs:200]: I prefer `for in` with `?` to `try_fold` with closure.

Comment (mu001999) [reply src/lib.rs]: Alright, but I think the two lines `fields.push(field);` can be written once outside the if-let block :)

Comment (TOETOE55) [reply src/lib.rs]: great

**Pull request merged**